doc/async-retrieval.md

```diff
@@ -0,0 +1,175 @@
+ ~ License, or (at your option) any later version.
+ ~ This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
+ ~ even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ ~ General Public License for more details.
+ ~ You should have received a copy of the GNU General Public License along with this program. If not,
+ ~ see <https://www.gnu.org/licenses/>.
+ -->
+
+# Asynchronous Data Retrieval
+
+The query service is able to retrieve data from an external data provider when it is missing from
+local storage. This could happen, for example, when a particular query node was disconnected from
+the network for some time and missed some blocks being finalized, or even when a node was connected
+but was not elected by HotShot to be on the DA committee, and thus did not receive the full data for
+some blocks.
+
+To support this, the availability data source -- the backend that powers the data availability API
+module -- is split into two parts. The **storage** layer implements local storage: a simple database
+that supports queries over the two main data structures (blocks and leaves) with various indexes. It
+represents a _best effort_ attempt to store and retrieve the data. Any query at any time is allowed
+to fail if data is missing. The **provider** layer implements retrieval of data from various
+external sources. Both of these layers are abstract, and can have various implementations. We
+currently support two implementations of storage -- backed by the local file system, or a Postgres
+database -- and one implementation of a provider, which fetches data from a peer query service.
+
+This document will focus on `FetchingDataSource`, which ties these two layers together to provide an
+infallible abstraction of HotShot data. Queries made via `FetchingDataSource` will never fail. If
+data is not available locally, the data source will transparently retrieve the data from a provider,
+and the request will block until the data becomes available.
+
+## Query Interface
+
+The query interface provided by `FetchingDataSource` is designed to support a variety of use cases:
+* Requests that should succeed or fail immediately based on whether the data is available locally
+* Requests that should block until data becomes available
+* Proactive retrieval of missing data to aid future requests
+
+The first two are enabled by the `Fetch` abstraction. `Fetch` is a special kind of future
+representing either data that has been retrieved already, or a pending attempt to fetch data. When
+awaited, `Fetch` will return immediately when the data is available, or block until it becomes
+available. But it is also possible to check if the data is available now and fail if not, using
+`Fetch::try_resolve()`.
+
+The two use cases of `Fetch` can be characterized as "on demand" fetching: we only look for data
```

Comment:
Seems confusing as there are actually three cases, the third of which is explained at the end of the paragraph. "*These* two use cases..." seems clearer to me.

Reply:
Ah, it was referring to the fact the only the first two of the 3 listed use cases are related to `Fetch`. I see how this is confusing, let me reword